src/pykeen/evaluation/evaluator.py

```diff
@@ -416,6 +385,74 @@ def _check_slicing_availability(model: Model, batch_size: int, entities: bool, r
             )
 
 
+def _hasher(kwargs: Mapping[str, Any]) -> int:
+    """Share optimal batch size whenever this hash matches."""
+    return hash((id(kwargs["evaluator"]), kwargs["mapped_triples"].shape[0], kwargs["targets"]))
+
+
+# note: we wrap this method twice; the inner one (=batch size) is reduced before the outer one applies
+@maximize_memory_utilization(hasher=_hasher, parameter_name="slice_size")
+@maximize_memory_utilization(hasher=_hasher)
+@torch.inference_mode()
+def optimized_evaluate(
+    *,
+    evaluator: Evaluator,
+    mapped_triples: MappedTriples,
+    batch_size: int,
+    slice_size: int,
+    progress_bar: tqdm,
+    targets: Collection[Target],
+    **kwargs,
+) -> MetricResults:
+    """
+    Evaluate a model with the given evaluator.
+
+    .. note ::
+        this method is wrapped into two memory utilization maximizer, which reduce the parameters `batch_size` and
+        `slice_size`, if necessary due to memory constraints.
+
+    :param evaluator:
+        the evaluator instance that is used for evaluation
+    :param mapped_triples:
+        the evaluation triples
+    :param batch_size:
+        the (maximum) batch size. It will be reduced when (GPU) out-of-memory problems occur.
+    :param slice_size:
+        the (maximum) slice size. It will be reduced when (GPU) out-of-memory problems occur, and batch size reduction
+        could not resolve the issue.
+    :param progress_bar:
+        whether to display a progress bar
+    :param targets:
+        the evaluation targets
+    :param kwargs:
+        additional keyword-based parameters are passed to :meth:`_evaluate_batch`.
+
+    :return:
+        the evaluation result
+    """
+    # todo: maybe we want to have some more keys outside of kwargs for hashing / have more visibility about
+    #  what is passed around
+    # clear evaluator and reset progress bar (necessary for size-probing / evaluation fallback)
+    evaluator.clear()
+    progress_bar.reset(total=mapped_triples.shape[0])
+
+    # batch-wise processing
+    for batch in mapped_triples.split(split_size=batch_size):
+        relation_filter = None
```

Comment:
can you add some text explaining what is happening to the relation filter on each iteration of `for target in targets`?

Reply:
@mberr can you take a look at this comment?

Reply:
oop i see it!